force-app/main/default/classes/DMLOperations.cls

```diff
@@ -146,11 +154,62 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
+        // Query the Account with the provided name
+        List<Account> accList = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+        
+        Account acc;
+        
+        if (!accList.isEmpty()) {
+            // Account with the given name exists, use it
+            acc = accList[0];
+        } else {
+            // Account with the given name doesn't exist, create a new one
+            acc = new Account(Name = accountName);
+            insert acc;
+        }
+    
+        // Retrieve existing Opportunities related to the Account
+        List<Opportunity> existingOpps = [SELECT Id, Name FROM Opportunity WHERE AccountId = :acc.Id];
+    
+        // Create a map to store existing Opportunity names
+        Map<String, Opportunity> existingOppsMap = new Map<String, Opportunity>();
+        for (Opportunity opp : existingOpps) {
+            existingOppsMap.put(opp.Name, opp);
+        }
+    
+        // Create a list to store Opportunity records for upsert
+        List<Opportunity> opportunities = new List<Opportunity>();
+    
         // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        for (String oppName : oppNames) {
+            // Check if an Opportunity with the same name already exists
+            if (existingOppsMap.containsKey(oppName)) {
+                // If exists, update the existing Opportunity
+                Opportunity existingOpp = existingOppsMap.get(oppName);
+                existingOpp.StageName ='Prospecting';
+                existingOpp.CloseDate = Date.today().addDays(30);
+                opportunities.add(existingOpp);
+            } else {
+                // If not exists, create a new Opportunity and associate it with the Account
+                Opportunity newOpp = new Opportunity(
+                    Name = oppName,
+                    AccountId = acc.Id,
+                    StageName ='Prospecting',
+                    CloseDate = Date.today().addDays(30)
+                );
+                opportunities.add(newOpp);
+            }
+        }
+    
+        // Perform upsert operation on the list of Opportunities
+        upsert opportunities;
+    
+        // Validate that the number of updated and newly inserted Opportunities matches the number of Opportunity names provided
+        System.assertEquals(oppNames.size(), opportunities.size(), 'The number of upserted Opportunities should match the number of Opportunity names provided');
```

Comment:
Great work with this method. You actually don't need an assert Statement here, and Assertions are generally not used in Apex outside of test classes.

(...there is a particular style of programming in some languages where you DO just make assertions as your code runs. Apex is not one of these languages, though.)